File: docs/validation/python/app/scenarios/s06_builder.py
import math
import tempfile
from machine_config import MachineConfigReader, MachineConfigWriter, MockConfigBuilder

"""
S-06: Build synthetic config with 2 lasers, verify fields and round-trip
ID:          S-06
Title:       Build a synthetic config with MockConfigBuilder and verify fields
Category:    happy-path
Layer:       builder
Precondition: None (builder creates from scratch)
Action:      Build a 2-laser config → verify fields → save to temp → re-read
Expected:    len(optical_trains) == 2
             train[0].scanner.scan_head_rotation ≈ 0.0
             train[1].scanner.scan_head_rotation ≈ 180.0
             machine_name non-empty
             correction_data centre cell ≈ 2.0 (Gaussian peak)
             After save and re-read: all fields match
Rationale:   Builder is used in CI fixture generation and by consumers who need
             synthetic test configs. Must produce valid, re-readable output.
"""

def run(fixtures_dir: str, real_dir: str) -> tuple[bool, str]:
    cfg = MockConfigBuilder(n_lasers=2).build()

    if len(cfg.optical_trains) != 2:
        return False, f"optical_trains count: {len(cfg.optical_trains)}"

    r0 = cfg.optical_trains[0].scanner.scan_head_rotation
    r1 = cfg.optical_trains[1].scanner.scan_head_rotation
    if r0 is None or abs(r0) > 0.001:
        return False, f"train[0].scan_head_rotation: {r0}"
    if r1 is None or abs(r1 - 180.0) > 0.001:
        return False, f"train[1].scan_head_rotation: {r1}"
    if not cfg.meta.machine_name:
        return False, "machine_name is empty"

    cb = cfg.optical_trains[0].optional_components.clearbox
    if cb is None or cb.correction_data is None:
        return False, "clearbox or correction_data is None"
    center = cb.correction_data[128][128][0]
    if not math.isfinite(center) or abs(center - 2.0) > 0.01:
        return False, f"correction_data center: expected ~2.0, got {center}"

    with tempfile.NamedTemporaryFile(suffix=".h5", delete=False) as f:
        tmp = f.name

    MachineConfigWriter(cfg).write(tmp)
    rb = MachineConfigReader(tmp).parse()

    if len(rb.optical_trains) != 2:
        return False, f"readback trains: {len(rb.optical_trains)}"
    if rb.meta.machine_name != cfg.meta.machine_name:
        return False, "machine_name changed after roundtrip"

    return True, "2-laser build OK, center≈2.0, roundtrip OK"

File: docs/validation/rust/app/src/scenarios/s06_builder.rs
// S-06: Build synthetic config with MockConfigBuilder and verify fields
//
// ID:          S-06
// Action:      Build a 2-laser config → verify fields → save to temp → re-read
// Expected:    2 trains, rotations 0/180, machine_name non-empty,
//              correction_data centre cell ≈ 2.0 (Gaussian peak), roundtrip OK.

use machine_config::{MachineConfigReader, MachineConfigWriter, MockConfigBuilder};
use std::path::Path;

pub fn run(_fixtures_dir: &Path, _real_dir: &Path) -> (bool, String) {
    let cfg = MockConfigBuilder::new(2).build();

    if cfg.optical_trains.len() != 2 {
        return (false, format!("optical_trains count: {}", cfg.optical_trains.len()));
    }

    let r0 = cfg.optical_trains[0].scanner.scan_head_rotation;
    let r1 = cfg.optical_trains[1].scanner.scan_head_rotation;
    match r0 {
        Some(r) if r.abs() <= 0.001 => {}
        other => return (false, format!("train[0].scan_head_rotation: {other:?}")),
    }
    match r1 {
        Some(r) if (r - 180.0).abs() <= 0.001 => {}
        other => return (false, format!("train[1].scan_head_rotation: {other:?}")),
    }
    if cfg.meta.machine_name.is_empty() {
        return (false, "machine_name is empty".to_string());
    }

    let clearbox = match &cfg.optical_trains[0].optional_components.clearbox {
        Some(cb) => cb,
        None => return (false, "clearbox is None".to_string()),
    };
    let grid = match &clearbox.correction_data {
        Some(g) => g,
        None => return (false, "correction_data is None".to_string()),
    };
    let center = match grid[128][128][0] {
        Some(v) => v,
        None => return (false, "correction_data centre cell is None".to_string()),
    };
    if !center.is_finite() || (center - 2.0).abs() > 0.01 {
        return (false, format!("correction_data center: expected ~2.0, got {center}"));
    }

    let tmp = match tempfile::Builder::new().suffix(".h5").tempfile() {
        Ok(t) => t,
        Err(e) => return (false, format!("tempfile creation failed: {e}")),
    };
    if let Err(e) = MachineConfigWriter::new(&cfg).write(tmp.path()) {
        return (false, format!("write failed: {e}"));
    }
    let rb = match MachineConfigReader::open(tmp.path()).and_then(|r| r.parse()) {
        Ok(c) => c,
        Err(e) => return (false, format!("readback failed: {e}")),
    };

    if rb.optical_trains.len() != 2 {
        return (false, format!("readback trains: {}", rb.optical_trains.len()));
    }
    if rb.meta.machine_name != cfg.meta.machine_name {
        return (false, "machine_name changed after roundtrip".to_string());
    }

    (true, "2-laser build OK, center≈2.0, roundtrip OK".to_string())
}
